get_routers/get_routers.py

```diff
@@ -0,0 +1,83 @@
+"""
+The following Python code sample shows how to access the
+"routers" endpoint of the NCM REST API using the
+Python "requests" module with paging. It makes use of NCM
+and CP credential headers to authenticate and access the
+router data. The router_id is the key and the router_group,
+router_name, and router_state are the values.
+"""
+
+import json
+import requests
+from datetime import datetime, timezone, timedelta
+
+headers = {
+    'X-CP-API-ID': '...',
+    'X-CP-API-KEY': '...',
+    'X-ECM-API-ID': '...',
+    'X-ECM-API-KEY': '...',
+    'Accept': '*/*',
+    'Content-Type': 'application/json',
+}
+
+
+account = ''  # your ncm account id here.
+base_url = 'https://www.cradlepointecm.com'
+
+
+def get_time():
+    """ Routers updated within the last 10 minutes. """
+    return (datetime.now(timezone.utc).astimezone()
+            - timedelta(minutes=10)).isoformat()
+
+
+def get_router_info(time):
+    """ Get router information """
+
+    url = base_url + f'/api/v2/routers/?limit=500&account={account}'
+
+    # we ultimately want to filter on state_updated_at to reduce api calls.
+    # url = base_url + f'/api/v2/routers/?limit=500&account={account} \
+    # &state_updated_at={time}'
+
+    routers = {}
+
+    while url:
+        req = requests.get(url, headers=headers)
```

Comment:
There is no error handling at all in this file.  At the very least, exceptions should be caught and handled gracefully.  Better, we should have retry code in here, since any http request across the internet could fail.  This script gives the (bad) impression that it is OK to assume no errors will occur.

Reply:
Here is what I came up with..

def get_router_info(now):
    """ Get router information """

    url = base_url + f'/api/v2/routers/?limit=500&account={account}\
    &state_updated_at__lt={now}'

    routers = {}

    while url:
        try:
            """ This will make all HTTP requests from the same session
            retry for a total of 3 times, sleeping between retries with an
            increasing backoff of 0s, 2s, 4s. It will retry on basic
            connectivity issues and the listed HTTP status codes. """

            session = requests.session()
            retries = Retry(total=3, backoff_factor=1,
                            status_forcelist=[ 408, 502, 503, 504 ])
            session.mount('https://', HTTPAdapter(max_retries=retries))
            response = session.get(url, headers=headers, timeout=3)

            if response.status_code == 401:
                print("Check your headers to make sure",
                      "you have the right credentials...")
                sys.exit(1)

            elif response.status_code == 200:
                # if you get 200/ok back, store the response in json format.
                resp = response.json()

        except requests.exceptions.HTTPError as errh:
            print("HTTP Error:",errh)

        except requests.exceptions.ConnectionError as errc:
            print("Connection Error:",errc)

        except requests.exceptions.Timeout as errt:
            print("Timeout Error:",errt)

        except requests.exceptions.RequestException as err:
            print("General Error:",err)

        if (len(resp['data']) < 1):
            return routers
        for items in resp['data']:

            if items['group']:
                group = items['group'].replace(
                    'https://www.cradlepointecm.com/api/v2/groups/', '')[:-1]
            else:
                group = None
            routers_dict = {'router_group': group,
                            'router_name': items['name'],
                            'router_state': items['state'], }

            router = {items['id']: routers_dict}
            routers.update(router)

            url = resp['meta']['next']

    return routers